substrate/primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -88,10 +89,96 @@ impl<V> Default for OverlayedEntry<V> {
 }
 
 /// History of value, with removal support.
-pub type OverlayedValue = OverlayedEntry<Option<StorageValue>>;
+pub type OverlayedValue = OverlayedEntry<StorageEntry>;
+
+/// Content in an overlay for a given transactional depth.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub enum StorageEntry {
+	/// A `set` operation was performed, overwrite previous
+	/// on commit or restore parent entry on rollback.
+	Some(StorageValue),
+	/// A `set` operation did remove value from the overlay.
+	None,
+	/// Contains the current appended value, number of item at start of transaction and offset at
+	/// start of transaction. A `append` operation did push content to the value, use previous
+	/// append info on commit or rollback by truncating to previous offset.
+	/// If a `set` operation occurs, store these to parent: overite on commit and restored on
+	/// rollback.
+	Append {
+		/// current buffer of appended data.
+		data: AppendData,
+		/// Current number of appended elements.
+		/// This is use to rewrite materialized size when needed.
+		nb_append: u32,
+		/// When define, contains the number of elements written in data as prefix.
+		/// If undefine, `data` do not contain the number of elements.
+		/// This number is updated on access only, it may differs from the actual `nb_append`.
+		materialized: Option<u32>,
```

Comment:
TODO could this be move to AppendData? (when moved we need it, when data present to, so only make sense if lazy fetching the initial value)